packages/babel-plugin-transform-react-constant-elements/src/index.js

```diff
@@ -75,6 +89,68 @@ export default declare((api, options) => {
         stop();
       }
     },
+    "ReferencedIdentifier|ThisExpression"(path, state) {
+      const { node } = path;
+      let { scope } = path;
+      let targetScope;
+
+      let isNestedScope = true;
+      let needsHoisting = true;
```

Comment:
I think we can actually reduce this code a bit:

```diff
diff --git a/packages/babel-plugin-transform-react-constant-elements/src/index.js b/packages/babel-plugin-transform-react-constant-elements/src/index.js
index 40d929e992..ccd6340f74 100644
--- a/packages/babel-plugin-transform-react-constant-elements/src/index.js
+++ b/packages/babel-plugin-transform-react-constant-elements/src/index.js
@@ -94,47 +94,11 @@ export default declare((api, options) => {
       let { scope } = path;
       let targetScope;
 
-      let isNestedScope = true;
-      let needsHoisting = true;
-
       while (scope) {
         // We cannot hoist outside of the previous hoisting target
         // scope, so we return early and we don't update it.
         if (scope === state.targetScope) return;
 
-        // When we hit the scope of our JSX element, we must start
-        // checking if they declare the binding of the current
-        // ReferencedIdentifier.
-        // We don't case about bindings declared in nested scopes,
-        // because the whole nested scope is hoisted alongside the
-        // JSX element so it doesn't impose any extra constraint.
-        if (scope === state.jsxScope) {
-          isNestedScope = false;
-        }
-
-        // If we are in an upper scope and hoisting to this scope has
-        // any benefit, we update the possible targetScope to the
-        // current one.
-        if (!isNestedScope && needsHoisting) {
-          targetScope = scope;
-        }
-
-        // When we start walking in upper scopes, avoid hoisting JSX
-        // elements until we hit a scope introduced by a function or
-        // loop.
-        // This is because hoisting from the inside to the outside
-        // of block or if statements doesn't give any performance
-        // benefit, and it just unnecessarily increases the code size.
-        if (scope === state.jsxScope) {
-          needsHoisting = false;
-        }
-        if (
-          !needsHoisting &&
-          (scope.path.isFunctionParent() || scope.path.isLoop())
-        ) {
-          needsHoisting = true;
-        }
-
         // If the current scope declares the ReferencedIdentifier we
         // are checking, we break out of this loop. There are two
         // possible scenarios:
@@ -148,8 +112,7 @@ export default declare((api, options) => {
 
         scope = scope.parent;
       }
-
-      if (targetScope) state.targetScope = targetScope;
+      if (scope) state.targetScope = scope.getFunctionParent();
     },
   };
 
@@ -189,12 +152,11 @@ export default declare((api, options) => {
           current = current.parentPath;
           jsxScope = HOISTED.get(current.node);
         }
-        jsxScope ??= path.scope;
+        jsxScope ??= path.scope.getFunctionParent();
 
         const state = {
           isImmutable: true,
           mutablePropsAllowed,
-          jsxScope,
           targetScope: null,
         };
 
```

Reply:
This unnecessarily caches some components:
```js
function Component({ isLoading, text }) {
  if (isLoading) {
    return <p>Loading {text}</p>;
  }
  return <p>Loaded {text}</p>;
}
```
```js
function Component({
  isLoading,
  text
}) {
  var _p;

  if (isLoading) {
    return _p || (_p = <p>Loading {text}</p>);
  }

  return <p>Loaded {text}</p>;
}
```
which is the reason I introduced the `needsHoisting` variable.

Also, I cannot drop `ReferencedIdentifier|ThisExpression` because it needs to handle things like `<p>{() => this}</p>`.

Reply:
I edited the diff a few hours after initially comment, are you sure you're using the latest? You can see it running at https://astexplorer.net/#/gist/15ad6fa36177a1ba8f06a94c7dd28f8a/bedfd7d67d72dbb98b33be1e66f98785c3b18aac. This test case (and all the others) are passing for me.

Reply:
I probably made some mistake while applying your diff.

I'm just commenting out my code instead of deleting it, since it will be needed if/when we do https://github.com/babel/babel/pull/12967#issuecomment-791352695 for cases like
```js
function Component({ isLoading, text }) {
  return () => <p>Loading {text => text}</p>;
}
```
which with the simplified logic would become
```js
function Component({
  isLoading,
  text
}) {
  return () => _p || (_p = <p>Loading {text => {
      var _p;

      return text;
    }}</p>);
}
```

Reply:
I think we'll be able to handle that case with the simplified code, but let's wait for the followup PR.